### DISCUSSION TITLE
Cannot gettype from within package

### DISCUSSION THREAD

**Discussion by finswimmer** [oracle/python-oracledb#181] [Q&A] [answered]
Hello,

using [gettype](https://python-oracledb.readthedocs.io/en/latest/api_manual/connection.html#Connection.gettype) works perfectly fine for global object types. But as soon as I'm trying to get a type from a package with `connection.gettype("<packagename>.<objecttype>")`  I receive an ORA-04043.

Are there any known limitation when working with oracle 11g?

I'm using:

* Python: 3.11
* python-oracledb: 1.3.1
* Oracle Version: 11.2.0.4.0 (output of `connection.version`)

fin swimmer

**Comment by cjbj** [comment, marked as answer]
What is possible for applications to access has definitely been improved in various Oracle DB versions. Look at the examples in https://github.com/oracle/python-oracledb/blob/main/tests/test_3200_features_12_1.py - which is the "Module for testing features introduced in 12.1". It shows various type cases were additionally supported.  Do these match what you are trying to do?

**Reply by finswimmer** [reply]
Yep, those tests look exactly like what I'm trying to do.  So I must assume, that's a limitation by Oracle 11g not to give access to types living in a package?

So the only workaround here seems to be, instead of passing the data directly to a procedure, I have to put them in a table and the procedure I'm calling must fetch the data from there. Or do you see any other chance here?

**Reply by cjbj** [reply]
A wrapper PL/SQL procedure or anonymous block will get you going.

> Or do you see any other chance here?

Upgrade!!!!

**Reply by mkhayom** [reply]
> A wrapper PL/SQL procedure or anonymous block will get you going.

Can you please give any example? 

i am testing example from [https://python-oracledb.readthedocs.io/en/latest/user_guide/bind.html#binding-pl-sql-records](docs )
 
I have successfully tested for global types, but code fails for types stored in package specification.


P.S. I use Oracle 11g and upgrading is not an option for my case.

**Comment by mkhayom** [comment]
```python 
with app_state.cft_pool.cursor() as cur:
        query = """
        declare
            INPUT Z$MYPKG.DEMORECORD;
        begin
            -- Call the procedure
            input.NUMBERVALUE := :num;
            input.STRINGVALUE := :str;
            Z$ELMA_USER_LIB_BI.TEST(INPUT => INPUT);
            :result := input.NUMBERVALUE || ' ' || input.STRINGVALUE;
        end;
        """
        res = cur.var(str)
        cur.execute(query, num=6, str="Test String", result=res)
        print(res.getvalue())
```

this worked well on Oracle 11g
